File: AndroidClient/LdpAndroidClient/app/src/main/java/com/ldp/androidclient/network/LdpPacketHandler.java
package com.ldp.androidclient.network;


import android.util.Log;

import com.ldp.androidclient.disposable.ILdpDisposable;
import com.ldp.androidclient.protocol.*;

import java.util.Iterator;
import java.util.concurrent.CopyOnWriteArrayList;

abstract public class LdpPacketHandler implements ILdpObservable, ILdpDisposable {
    private static final String TAG = "LdpPacketHandler";

    protected abstract void handle();

    private CopyOnWriteArrayList<ILdpPacketHandler> listeners;

    public LdpPacketHandler() {
        listeners = new CopyOnWriteArrayList<ILdpPacketHandler>();
    }

    @Override
    public void addListener(ILdpPacketHandler listener) {
        if (!listeners.contains(listener))
            listeners.add(listener);
    }

    @Override
    public void notifyToAllListeners(LdpProtocol.LdpPacket packet) {
        if (listeners != null && listeners.size() != 0) {
            Iterator<ILdpPacketHandler> iterator = listeners.iterator();
            while (iterator.hasNext()) {
                ILdpPacketHandler lis = iterator.next();
                lis.handle(packet);
            }
        }
    }

    @Override
    public void removeListener(ILdpPacketHandler listener) {
        if (listeners.contains(listener)) {
            listener.dispose();
            listeners.remove(listener);
            Log.i(TAG, "removed listener: " + listener.toString());
        }
    }

    @Override
    public void removeListeners() {
        try {
            if (listeners != null && listeners.size() != 0) {
                Iterator<ILdpPacketHandler> iterator = listeners.iterator();
                while (iterator.hasNext()) {
                    ILdpPacketHandler listener = iterator.next();
                    listener.dispose();
                    listener = null;
                }
                if (listeners != null)
                    listeners.clear();
                listeners = null;
            }
        } catch (Exception ignored) {

        }

    }

    @Override
    public void dispose() {
        removeListeners();
    }
}

File: Server/Server/Network/Handlers/PacketHandlerBase/LdpPacketHandler.cs
using Server.Protocol;
using Server.WindowsUtils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Server.Network.Handlers.PacketHandlerBase
{
    abstract class LdpPacketHandler : ILdpObservable, IDisposable
    {
        protected abstract void Handle();

        private List<ILdpPacketHandler> listeners;
        public LdpPacketHandler()
        {
            listeners = new List<ILdpPacketHandler>();
        }

        #region Observer pattern impl
        public void AddListener(ILdpPacketHandler listener)
        {
            if (!listeners.Contains(listener))
                listeners.Add(listener);
        }

        public void RemoveListener(ILdpPacketHandler listener)
        {
            if (listeners.Contains(listener))
            {
                listener.Dispose();
                listeners.Remove(listener);
                LdpLog.Info(String.Format("Listener removed: {0}.", listener.GetType()));
                listener = null;
            }
        }

        public void RemoveListeners()
        {
            if (listeners != null && listeners.Count != 0)
            {
                for (int i = listeners.Count - 1; i >= 0; i--)
                {
                    LdpLog.Info(String.Format("Listener removed: {0}.",
                        listeners[i].GetType()));
                    listeners[i].Dispose();
                    listeners[i] = null;
                    listeners.Remove(listeners[i]);
                }
            }
        }

        public void NotifyToAllListeners(LdpPacket packet)
        {
            if (listeners != null && listeners.Count != 0)
            {
                foreach (var listener in listeners.ToList())
                {
                    ILdpPacketHandler ipl = (ILdpPacketHandler)listener;
                    ipl.Handle(packet);
                }
            }
        }
        #endregion


        public List<ILdpPacketHandler> GetListenersList
        {
            get 
            {
                if (listeners != null && listeners.Count != 0)
                    return listeners;
                else return new List<ILdpPacketHandler>();
            }
        }

        public virtual void Dispose()
        {
            RemoveListeners();
            listeners = null;
            GC.SuppressFinalize(this);
        }
    }
}
